File: eip_toolkit_ultra_simple.py
#!/usr/bin/env python3
"""
EIP Monitor - Ultra Simple Version
Monitor Azure Red Hat OpenShift Egress IP assignments.
"""

import json
import os
import subprocess
import sys
import time
from datetime import datetime


def run_oc(cmd):
    """Run oc command and return JSON."""
    result = subprocess.run(['oc'] + cmd, capture_output=True, text=True, check=True)
    return json.loads(result.stdout)


def run_az(cmd):
    """Run az command and return JSON."""
    subscription = os.getenv('AZ_SUBSCRIPTION')
    if not subscription:
        raise Exception("AZ_SUBSCRIPTION environment variable not set")
    
    result = subprocess.run(['az'] + cmd + ['--subscription', subscription], 
                          capture_output=True, text=True, check=True)
    return json.loads(result.stdout)


def get_eip_stats():
    """Get EIP statistics."""
    data = run_oc(['get', 'eip', '-o', 'json'])
    assigned = len([item for item in data['items'] if item.get('status', {}).get('items', [])])
    configured = len(data['items'])
    return assigned, configured


def get_cpic_stats():
    """Get CPIC statistics."""
    data = run_oc(['get', 'cloudprivateipconfig', '-o', 'json'])
    success = len([item for item in data['items'] 
                  for condition in item.get('status', {}).get('conditions', [])
                  if condition.get('reason') == 'CloudResponseSuccess'])
    return success


def get_nodes():
    """Get EIP-enabled nodes."""
    result = subprocess.run(['oc', 'get', 'nodes', '-l', 'k8s.ovn.org/egress-assignable=true', '-o', 'name'],
                          capture_output=True, text=True, check=True)
    return [line.replace('node/', '') for line in result.stdout.strip().split('\n') if line]


def get_azure_nic_stats(node):
    """Get Azure NIC statistics for a node."""
    try:
        nic_name = f"{node}-nic"
        resource_group = os.getenv('AZ_RESOURCE_GROUP')
        if not resource_group:
            return 0, 0
        
        # Get IP configurations
        ip_configs = run_az(['network', 'nic', 'show', '--resource-group', resource_group, 
                           '--name', nic_name, '--query', 'ipConfigurations[].privateIPAddress'])
        
        # Get load balancer associations
        lb_configs = run_az(['network', 'nic', 'show', '--resource-group', resource_group,
                           '--name', nic_name, '--query', 'ipConfigurations[].{pools:loadBalancerBackendAddressPools[].id}'])
        
        total_ips = len(ip_configs) if ip_configs else 0
        lb_associated = len([cfg for cfg in lb_configs if cfg.get('pools')]) if lb_configs else 0
        
        # Subtract primary IP
        secondary_ips = max(0, total_ips - 1)
        secondary_lb_ips = max(0, lb_associated - 1)
        
        return secondary_ips, secondary_lb_ips
    except:
        return 0, 0


def monitor():
    """Main monitoring loop."""
    print("Starting EIP monitoring...")
    
    nodes = get_nodes()
    print(f"Found EIP-enabled nodes: {nodes}")
    
    while True:
        timestamp = datetime.now().strftime('%H:%M:%S')
        
        # Get global stats
        assigned, configured = get_eip_stats()
        cpic_success = get_cpic_stats()
        
        # Get Azure stats for each node
        azure_stats = []
        for node in nodes:
            azure_eips, azure_lbs = get_azure_nic_stats(node)
            azure_stats.append(f"{node}:{azure_eips}/{azure_lbs}")
        
        print(f"[{timestamp}] EIPs: {assigned}/{configured}, CPIC: {cpic_success}, Azure: {', '.join(azure_stats)}")
        
        # Check if complete
        if assigned == configured and cpic_success == configured:
            print("✅ All EIPs assigned and CPIC issues resolved!")
            break
        
        time.sleep(1)


if __name__ == '__main__':
    try:
        monitor()
    except KeyboardInterrupt:
        print("\nMonitoring stopped by user")
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)

File: eip_toolkit_ultra_simple.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"strings"
	"time"
)

// runOC executes oc command and returns JSON result
func runOC(cmd []string) (map[string]interface{}, error) {
	args := append([]string{}, cmd...)
	cmdObj := exec.Command("oc", args...)
	output, err := cmdObj.Output()
	if err != nil {
		return nil, fmt.Errorf("oc command failed: %w", err)
	}

	var result map[string]interface{}
	if err := json.Unmarshal(output, &result); err != nil {
		return nil, fmt.Errorf("failed to parse JSON: %w", err)
	}

	return result, nil
}

// runAZ executes az command and returns JSON result
func runAZ(cmd []string) (interface{}, error) {
	subscription := os.Getenv("AZ_SUBSCRIPTION")
	if subscription == "" {
		return nil, fmt.Errorf("AZ_SUBSCRIPTION environment variable not set")
	}

	args := append([]string{}, cmd...)
	args = append(args, "--subscription", subscription)
	cmdObj := exec.Command("az", args...)
	output, err := cmdObj.Output()
	if err != nil {
		return nil, fmt.Errorf("az command failed: %w", err)
	}

	var result interface{}
	if err := json.Unmarshal(output, &result); err != nil {
		return nil, fmt.Errorf("failed to parse JSON: %w", err)
	}

	return result, nil
}

// getEIPStats returns assigned and configured EIP counts
func getEIPStats() (int, int, error) {
	data, err := runOC([]string{"get", "eip", "--all-namespaces", "-o", "json"})
	if err != nil {
		return 0, 0, err
	}

	items, ok := data["items"].([]interface{})
	if !ok {
		return 0, 0, fmt.Errorf("invalid items format")
	}

	// Count configured EIPs: sum up the number of IPs in spec.egressIPs for each EIP resource
	// Each EIP resource can have multiple IPs configured (e.g., 2 IPs per namespace)
	configured := 0
	assigned := 0

	for _, item := range items {
		itemMap, ok := item.(map[string]interface{})
		if !ok {
			continue
		}

		// Count IPs in spec.egressIPs
		spec, ok := itemMap["spec"].(map[string]interface{})
		if ok {
			if egressIPs, ok := spec["egressIPs"].([]interface{}); ok {
				configured += len(egressIPs)
			}
		}

		// Count assigned IPs from status.items
		status, ok := itemMap["status"].(map[string]interface{})
		if !ok {
			continue
		}

		statusItems, ok := status["items"].([]interface{})
		if ok {
			assigned += len(statusItems)
		}
	}

	return assigned, configured, nil
}

// getCPICStats returns successful CPIC count
func getCPICStats() (int, error) {
	data, err := runOC([]string{"get", "cloudprivateipconfig", "--all-namespaces", "-o", "json"})
	if err != nil {
		return 0, err
	}

	items, ok := data["items"].([]interface{})
	if !ok {
		return 0, fmt.Errorf("invalid items format")
	}

	success := 0

	for _, item := range items {
		itemMap, ok := item.(map[string]interface{})
		if !ok {
			continue
		}

		status, ok := itemMap["status"].(map[string]interface{})
		if !ok {
			continue
		}

		conditions, ok := status["conditions"].([]interface{})
		if !ok {
			continue
		}

		for _, cond := range conditions {
			condMap, ok := cond.(map[string]interface{})
			if !ok {
				continue
			}

			reason, ok := condMap["reason"].(string)
			if ok && reason == "CloudResponseSuccess" {
				success++
				break
			}
		}
	}

	return success, nil
}

// getNodes returns list of EIP-enabled nodes
func getNodes() ([]string, error) {
	cmd := exec.Command("oc", "get", "nodes", "-l", "k8s.ovn.org/egress-assignable=true", "-o", "name")
	output, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("cannot access OpenShift nodes: %w", err)
	}

	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	nodes := make([]string, 0, len(lines))

	for _, line := range lines {
		if line == "" {
			continue
		}
		nodeName := strings.TrimPrefix(line, "node/")
		if nodeName != "" {
			nodes = append(nodes, nodeName)
		}
	}

	return nodes, nil
}

// getAzureNICStats returns secondary IPs and load balancer IPs for a node
func getAzureNICStats(node string) (int, int) {
	resourceGroup := os.Getenv("AZ_RESOURCE_GROUP")
	if resourceGroup == "" {
		return 0, 0
	}

	nicName := fmt.Sprintf("%s-nic", node)

	// Get IP configurations
	ipConfigsResult, err := runAZ([]string{"network", "nic", "show",
		"--resource-group", resourceGroup,
		"--name", nicName,
		"--query", "ipConfigurations[].privateIPAddress"})
	if err != nil {
		return 0, 0
	}

	ipConfigs, ok := ipConfigsResult.([]interface{})
	if !ok {
		return 0, 0
	}

	// Get load balancer associations
	lbConfigsResult, err := runAZ([]string{"network", "nic", "show",
		"--resource-group", resourceGroup,
		"--name", nicName,
		"--query", "ipConfigurations[].{pools:loadBalancerBackendAddressPools[].id}"})
	if err != nil {
		return len(ipConfigs) - 1, 0
	}

	lbConfigs, ok := lbConfigsResult.([]interface{})
	if !ok {
		return max(0, len(ipConfigs)-1), 0
	}

	totalIPs := len(ipConfigs)
	lbAssociated := 0

	for _, cfg := range lbConfigs {
		cfgMap, ok := cfg.(map[string]interface{})
		if !ok {
			continue
		}

		pools, ok := cfgMap["pools"].([]interface{})
		if ok && len(pools) > 0 {
			lbAssociated++
		}
	}

	secondaryIPs := max(0, totalIPs-1)
	secondaryLBIPs := max(0, lbAssociated-1)

	return secondaryIPs, secondaryLBIPs
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// monitor runs the main monitoring loop
func monitor() error {
	fmt.Println("Starting EIP monitoring...")

	nodes, err := getNodes()
	if err != nil {
		return err
	}
	fmt.Printf("Found EIP-enabled nodes: %v\n", nodes)

	for {
		timestamp := time.Now().Format("15:04:05")

		// Get global stats
		assigned, configured, err := getEIPStats()
		if err != nil {
			return err
		}

		cpicSuccess, err := getCPICStats()
		if err != nil {
			return err
		}

		// Get Azure stats for each node
		var azureStats []string
		for _, node := range nodes {
			azureEIPs, azureLBs := getAzureNICStats(node)
			azureStats = append(azureStats, fmt.Sprintf("%s:%d/%d", node, azureEIPs, azureLBs))
		}

		fmt.Printf("[%s] EIPs: %d/%d, CPIC: %d, Azure: %s\n",
			timestamp, assigned, configured, cpicSuccess, strings.Join(azureStats, ", "))

		// Check if complete
		if assigned == configured && cpicSuccess == configured {
			fmt.Println("✅ All EIPs assigned and CPIC issues resolved!")
			break
		}

		time.Sleep(1 * time.Second)
	}

	return nil
}

func main() {
	if err := monitor(); err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
}

